app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/EnqueuedView.kt

```diff
@@ -1,77 +1,219 @@
-			return arrayOf(checkmark,
-						"",
-						UnicodeCleaner.clean(item.title ?: ""))
+
+			val coverArtImage = if (item.coverArt != null) graphicsHelpers.compress(item.coverArt!!, 90, 90, quality = 30) else ""
+
+			var title = UnicodeCleaner.clean(item.title ?: "")
+			if (title.length > ROW_LINE_MAX_LENGTH) {
+				title = title.substring(0, 20) + "..."
+			}
+			val artist = UnicodeCleaner.clean(item.artist ?: "")
+			val songMetaDataText = "${title}\n${artist}"
+
+			return arrayOf(
+					checkmark,
+					coverArtImage,
+					"",
+					songMetaDataText
+			)
 		}
 	}
 
 	init {
 		state as RHMIState.PlainState
 
 		listComponent = state.componentsList.filterIsInstance<RHMIComponent.List>().first()
+		queueImageComponent = state.componentsList.filterIsInstance<RHMIComponent.Image>().first()
+		titleLabelComponent = state.componentsList.filterIsInstance<RHMIComponent.Label>()[0]
+		subtitleLabelComponent = state.componentsList.filterIsInstance<RHMIComponent.Label>()[1]
 
 		songsEmptyList.addRow(arrayOf("", "", L.MUSIC_QUEUE_EMPTY))
 	}
 
 	fun initWidgets(playbackView: PlaybackView) {
-		state.getTextModel()?.asRaDataModel()?.value = L.MUSIC_QUEUE_TITLE
+		// this is required for pagination system
+		listComponent.setProperty(RHMIProperty.PropertyId.VALID, false)
+
 		listComponent.setVisible(true)
-		listComponent.setProperty(RHMIProperty.PropertyId.LIST_COLUMNWIDTH, "57,50,*")
+		listComponent.setProperty(RHMIProperty.PropertyId.LIST_COLUMNWIDTH, "57,90,10,*")
 		listComponent.getAction()?.asHMIAction()?.getTargetModel()?.asRaIntModel()?.value = playbackView.state.id
 		listComponent.getAction()?.asRAAction()?.rhmiActionCallback = RHMIActionListCallback { onClick(it) }
+		listComponent.getSelectAction()?.asRAAction()?.rhmiActionCallback = RHMIActionListCallback { onSelectAction(it) }
 	}
 
 	fun show() {
 		currentSong = musicController.getMetadata()
+		val newQueueMetadata = musicController.getQueue()
+
+		// same queue as before, just select currently playing song
+		if (queueMetadata != null && queueMetadata == newQueueMetadata) {
+			showCurrentlyPlayingSong()
+			return
+		}
+
+		queueMetadata = newQueueMetadata
 		songsList.clear()
-		val songs = musicController.getQueue()
+		val songs = queueMetadata?.songs
 		if (songs?.isNotEmpty() == true) {
 			listComponent.setEnabled(true)
 			listComponent.setSelectable(true)
 			songsList.addAll(songs)
-			listComponent.getModel()?.setValue(songsListAdapter, 0, songsListAdapter.height, songsListAdapter.height)
-
-			// set the selection to the current song
-			val index = songs.indexOfFirst { it.queueId == currentSong?.queueId }
-			if (index >= 0) {
-				state.app.events.values.firstOrNull { it is RHMIEvent.FocusEvent }?.triggerEvent(
-						mapOf(0.toByte() to listComponent.id, 41.toByte() to index)
-				)
+
+			listComponent.requestDataCallback = RequestDataCallback { startIndex, numRows ->
```

Comment:
I'm not sure if the requestDataCallback is triggered if the user revisits an existing row that was already drawn... could `visibleRows` be updated based on the selectAction instead of the RequestDataCallback?

Reply:
However, redrawing naively based on `selectAction` would also be suboptimal! Hmmmm
The edge case I have in mind (which seems rare, I don't think I've managed to trigger it yet) is:

1. RequestDataCallback requests data from 0-39, which gets set as the visibleRows
2. SpotifyAppController starts filling in the cover art for 0-39, but only makes it to, say, 10 or so
3. The redraw cycle updates the 0-10 items with the cover art
3. RequestDataCallback requests data from 40-79, which is the new visibleRows
4. SpotifyAppController continues filling in the cover art from 11-79
5. The redraw cycle waits and then starts filling in cover arts from 40 and on
6. The user scrolls back to the top of the list, but only 0-10 items have cover art
7. No new RequestDataCallback for 0-39 fires again, so the visibleRows never gets set back to the top of the list and it doesn't redraw to update the coverarts up there.

Do I grok the algorithm correctly?
One solution that comes to my mind is to track an array of which items (perhaps just tracked by index) have had their coverarts posted to the car, and only redraw the visibleRows if any of the coverarts are missing from that window.

What are your thoughts? You probably know better about the edges of your algorithm, especially since you've been testing it far longer than I :)

Reply:
Hmm that's an interesting case but thankfully that should never happen. The algorithm you have is slightly incorrect mainly around the times/how the requestDataCallback is called. Here's roughly how the pattern of requestDataCallback method is called: 

- Assume starting from index 0. 
- requestDataCallback is called for indexes 0 to 40
- user scrolls to index 30 --> requestDataCallback is called for indexes 40-50
- user scrolls to index 40 --> requestDataCallback is called for indexes 50-60
- user scrolls back to index 29 --> requestDataCallback is called for indexes 9-20
- user scrolls to index 39 --> requestDataCallback is called for indexes 49-59

So essentially the idea is that as long as we are storing a max of 41 items in the LRU the full range of the track cover arts should be supported at any one time and visible "window" will be shifted appropriately, even past the items that were previously viewed. I double checked this logic over and a scenario like this should never occur unless the car does something unpredictable :).